mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,21 +139,31 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
-
-      // Compute threshold to get around k elements.
-      // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
-      // so we pick quantile p = M / N + err
-      // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
-      val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
       val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
-      if (approxQuantile >= 1) {
-        modelDatasetWithDist
+      // for a small dataset, use BoundedPriorityQueue
+      if (count < 1000) {
+        val queue = new BoundedPriorityQueue[Double](count.toInt)(Ordering[Double])
```

Comment:
This place should be like:
```scala
val exactThreshold = modelDatasetWithDist
.select(distCol)
.as[Double]
.rdd
.treeAggregate(new BoundedPriorityQueue[Double](numNearestNeighbors)(Ordering[Double].reverse))(
seqOp= (q, v) => q += v,
combOp = (q1, q2) => q1 ++= q2,
depth = 2
).toArray.max
```

And this impl should have no dependency on the size of dataset.

Reply:
this only depends on `numNearestNeighbors`, when it is small (maybe < 10000?).
On each partition, collect the minmum 10 values, and merge them by `treeAggregate` to get the global minmum 10 values, and the max value in them is the threshold.

Reply:
`BoundedPriorityQueue` only maintains the topK entries, so it is safe to absorb a lot of entries.

Reply:
approxQuantile already kind of works this way, so I think the point of this PR is avoiding several passes of Spark jobs for tree reduce in this case.

However it's a fair point, I wonder if, overall, this approach is faster than approxQuantile? it already does something like what you're suggesting.

Reply:
I tried a small test. I used existing ```BucketedRandomProjectionLSHSuite``` but made the dataset bigger: 
```
val data = {
      for (i <- -200 until 200; j <- -200 until 200) yield Vectors.dense(i*10, j*10)
}
dataset = spark.createDataFrame(data1.map(Tuple1.apply)).toDF("keys")
```
So the dataset count is 160000 and I tested 10000, 9000, 8000, 7000, 6000, 5000, 4000, 3000, 2000 and 1000 nearest neighbors. I didn't see any performance gain using ```BoundedPriorityQueue```.

Reply:
OK, so we're back to the original question: is it worth collect and top-k via priority queue for a small data set? I'm OK with it, if it speeds things up. Are there any rough benchmarks for small data sets like 10 or 500 elements? the additional code complexity isn't too bad.

Reply:
I wrongly thought that the `approxNearestNeighbors` only return an approximate threshold, then we can use top-k to obtain an exact threshold.
Since the `approxNearestNeighbors` already gaurantee an enough threshold which had already taken the relative error into account, so **I guess we no longer need a top-k solution.**
A  `BoundedPriorityQueue` only maintains the topK entries, so it should be much smaller than a `QuantileSummaries`, however since there is only one column to process, so there should be no performance gain.

Reply:
I wrongly thought that the `approxNearestNeighbors` only return an approximate threshold, then we can use top-k to obtain an exact threshold.
Since the `approxNearestNeighbors` already gaurantee an enough threshold which had already taken the relative error into account, so **I guess we no longer need a top-k solution.**
A  `BoundedPriorityQueue` only maintains the topK entries, so it should be much smaller than a `QuantileSummaries`, however since there is only one column to process, so there should be no performance gain.

Reply:
A slight performance gain may come from that ` BoundedPriorityQueue` do not need a `count` job to compute the var `approxQuantile`.

Reply:
I agree just implementing this in terms of BoundedPriorityQueue is also a viable solution. I don't know which one is faster. I had assumed the approximate quantile would be as I think it does less work overall, but I haven't tested it. That is, I think it will hang on to fewer than k entries per partition

Reply:
I am trying to do some performance tests, however I found that in the public `approxNearestNeighbors` methods , `singleProbe` is always true for now, so our changes in LSH can not be reached in the code path.